### PR TITLE
Basic refractiveindex.info importer

### LINKED ISSUE #67: Data loader for refractiveindex.info
To supply many tabulated materials it would be helpful to have a loader for refractive index.info data.

For now we could just work with a plugin which takes the url of a csv file loading it into a tabular dispersion. Later I could imagine that we even could have some form of material search that one just has to type in the materials name.

### PR DESCRIPTION
**Pull request by MarJMue**
Fixes #67 

Still to do:
- [x] think about a unique identifier, because the index number changes with updates
- [x] implement dispersion formulas used in the database
- [x] helper functions to search the database
- [x] write tests
- [x] Documentation 
- [x] think about a way to distribute the database (internally/externally/download?)

### CODE REVIEW

**Comment by domna** [comment]
To integrate the database I thought we could use git submodules. So it would be directly available. Maybe it's also possible to add it via an extra requirement in setup.cfg.

I think in the [related projects](https://github.com/polyanskiy/refractiveindex.info-database#related-projects), there are also some projects which already handle searching in the database (Probably you already looked at them ,though). So maybe we can use some of their code.

**Comment by MarJMue** [comment]
I was not sure if we want to increase the size of our module by ~30 Mb. Package separately or include directly? Downloading on demand is no option from my point of view, because i do not like this kind of behaviour. 

pyTMM just assumes knowledge of the database structure.
Tthe other project is a full sql database implementation, maybe we should adapt this module, because is has approximate searching already implemented. I have to look into the code how we can fit it into our module.
Alternatively i thought about writing simple search functions to filter the dataframe, but i would need to do a bit more work to get it nice.

**Comment by MarJMue** [comment]
@domna i think i got most of the basic functionality and some rudamentary unit tests in place. If you want, you can take a look if you like the general implementation. If you approve, i will do some more cleanup and documention.

**Comment by domna** [comment]
This looks very nice!

One small suggestions, though:
I thought it may be simpler to just have two parameters in load_dispersions instead of a tuple. The second parameter could also be optional, so that the first one or the widest range etc is selected (I don't know if there is a default flag in the database, which would be the one which is preselected on the website, but maybe this is just the first one...).

I did not test the fuzzy search yet but I will play around w/ it the next days.

Edit: Just had an additional thought. Do you think we should place the file in a separate folder, e.g. db? I could imagine that we want to extend the functionality of searching and data handling on the database. I also thought `import db.DatabaseRII` or even `import db.RII` would be a bit more self explanatory in the api. I also could think of that we extend to other databases at some point or give the opportunity to have a local self-managed database (if I at some point I would decide to work further on the pyRelli REST api for example).

**Comment by MarJMue** [comment]
Yes will change the tuple thing. 
The database provides no real tags or anything. But selecting the first one might be a good idea. 

The fuzzy search works well enougth, we might tweak the default values a bit...
I'm just a bit unhappy with the way it is selected. maybe another method would be better.

The range question might be one to consider... Maybe i should implement a warning if a model is used outside of its range.

+1 for the db submodule.

**Comment by domna** [comment]
Yeah, I think the range question is a general one we should consider for dispersions. I think rn it just throws an interpolation error for table, we could just re-raise that in a meaningful way. For formulas it should be generally fine, though. Or do we have models with a range restriction? (Except Tauc-Lorentz where I'm not sure if we properly catch E>Eg). But maybe we should solve that in a separate issue?

For the fuzzy search I'm not sure as well. Maybe it is nice to set it permanently, because I think people either will use it or not. We could also think about just using fuzzy every time and throwing everything at the user (which is how searches in OSs etc. typically work, right?). 

What would be extremely nice (but this is just a feature idea) would be that the search display is selectable, such that you can just select the material and get it right away. Could be supported by ncurses in the terminal and some ipywidget for jlab maybe (dunno if there exists something for it already).

**Comment by MarJMue** [comment]
On the topic of a db submodule, as elli.AIR has quite a special case... do we want to move it too (elli.db.AIR)?

> What would be extremely nice (but this is just a feature idea) would be that the search display is selectable, such that you can just select the material and get it right away. Could be supported by ncurses in the terminal and some ipywidget for jlab maybe (dunno if there exists something for it already).

I like the simplicity of the current solution and think we should focus on TUI/GUI separately and for the complete workflow.

**Comment by domna** [comment]
I think it would only make sense if we could also access other materials as elli.db.material. But we cannot, right? So probably it's easiest to just keep it under elli.AIR?

> I like the simplicity of the current solution and think we should focus on TUI/GUI separately and for the complete workflow.

Yes, makes sense to approach it in an holistic approach 👍️

**Comment by MarJMue** [comment]
> I think it would only make sense if we could also access other materials as elli.db.material. But we cannot, right? So probably it's easiest to just keep it under elli.AIR?

It would be very nice to access the database this way but not possible, without manually parsing the whole database (as far as i can tell. We can access file with the dot notation, but as they are yaml and not python there is no code to execute.)
I reckon we could manually add useful shortcuts to specific entries of the RII database (e.g the Aspnes' Silicon)

**Comment by MarJMue** [comment]
I think i have done everything needed for the basic functionality.
Only question is if we want to revert the elli.AIR commit.

File: src/elli/dispersions/base_dispersion.py [hunk 1]
```diff
@@ -224,7 +224,7 @@ def __init__(self, *disps: Dispersion) -> None:
         self.dispersions = disps
 
     def dielectric_function(self, lbda: npt.ArrayLike) -> npt.NDArray:
-        dielectric_function = np.sum(
+        dielectric_function = sum(
```
Comment (domna) [inline src/elli/dispersions/base_dispersion.py:227]: Why not use numpy's sum? Should be faster as these are all numpy arrays, right?

**Review by domna — APPROVED** [review]

**Comment by domna** [comment]
I would find it nice if we use just the shorter notation of `elli.AIR`. We could also supply both to keep it sorted, when we also offer other materials through elli.db.

If we stay with `elli.db.AIR` the `generate_ipynb_from_gallery.sh` needs to be run to update the jupyter example files, too.

File: src/elli/dispersions/base_dispersion.py (hunk 1 above)
Comment (MarJMue) [reply src/elli/dispersions/base_dispersion.py:227]: There is a deprecation warning and i can't get the numpy solution to work. I'll add it as an separate issue.

> DeprecationWarning: Calling np.sum(generator) is deprecated, and in the future will give a different result. Use np.sum(np.fromiter(generator)) or the python sum builtin instead.

Comment (domna) [reply src/elli/dispersions/base_dispersion.py:227]: I see. I wasn't aware of this. I don't know if the numpy version gives so much benefit over just using sum here, so probably it is not worth the effort of digging deeper.

**Pull request merged**